avl_tree/avl_tree.py

```diff
@@ -1,81 +1,222 @@
+#     self.balance = 0
+
+#   """
+#   Display the whole tree. Uses recursive def.
+#   """
+#   def display(self, level=0, pref=''):
+#     self.update_height()  # Update height before balancing
+#     self.update_balance()
     
-    if self.node != None: 
-      print ('-' * level * 2, pref, self.node.key,
-        f'[{self.height}:{self.balance}]',
-        'L' if self.height == 0 else ' ')
-      if self.node.left != None:
-        self.node.left.display(level + 1, '<')
-      if self.node.right != None:
-        self.node.right.display(level + 1, '>')
-
-  """
-  Computes the maximum number of levels there are
-  in the tree
-  """
-  def update_height(self):
-    pass
-
-  """
-  Updates the balance factor on the AVLTree class
-  """
-  def update_balance(self):
-    pass
-
-  """
-  Perform a left rotation, making the right child of this
-  node the parent and making the old parent the left child
-  of the new parent. 
-  """
-  def left_rotate(self):
-    pass
-
-  """
-  Perform a right rotation, making the left child of this
-  node the parent and making the old parent the right child
-  of the new parent. 
-  """
-  def right_rotate(self):
-    pass
-
-  """
-  Sets in motion the rebalancing logic to ensure the
-  tree is balanced such that the balance factor is
-  1 or -1
-  """
-  def rebalance(self):
-    pass
+#     if self.node != None: 
+#       print ('-' * level * 2, pref, self.node.key,
+#         f'[{self.height}:{self.balance}]',
+#         'L' if self.height == 0 else ' ')
+#       if self.node.left != None:
+#         self.node.left.display(level + 1, '<')
+#       if self.node.right != None:
+#         self.node.right.display(level + 1, '>')
+
+#   """
+#   Computes the maximum number of levels there are
+#   in the tree
+#   """
+#   def update_height(self):
+#     pass
+
+#   """
+#   Updates the balance factor on the AVLTree class
+#   """
+#   def update_balance(self):
+#     pass
+
+#   """
+#   Perform a left rotation, making the right child of this
+#   node the parent and making the old parent the left child
+#   of the new parent. 
+#   """
+#   def left_rotate(self):
+#     pass
+
+#   """
+#   Perform a right rotation, making the left child of this
+#   node the parent and making the old parent the right child
+#   of the new parent. 
+#   """
+#   def right_rotate(self):
+#     pass
+
+#   """
+#   Sets in motion the rebalancing logic to ensure the
+#   tree is balanced such that the balance factor is
+#   1 or -1
+#   """
+#   def rebalance(self):
+#     pass
     
-  """
-  Uses the same insertion logic as a binary search tree
-  after the value is inserted, we need to check to see
-  if we need to rebalance
-  """
-  def insert(self, key):
-    pass
+#   """
+#   Uses the same insertion logic as a binary search tree
+#   after the value is inserted, we need to check to see
+#   if we need to rebalance
+#   """
+#   def insert(self, key):
+#     pass
+
+# Python code to insert a node in AVL tree 
+
```

Comment:
oooh, where did you get this code? i'm glad you have working implementation; if you had to take it from somewhere, i'm happy that you at least made an attempt to understand it!